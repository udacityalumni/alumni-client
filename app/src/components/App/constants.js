export const APP_SET_MOBILE = 'APP_SET_MOBILE';
export const APP_ON_TOGGLE_NAV = 'APP_ON_TOGGLE_NAV';
export const SET_SEARCH_TERM = 'SET_SEARCH_TERM';
export const CLEAR_SEARCH_TERM = 'CLEAR_SEARCH_TERM';
export const SET_AUTH_USER = 'SET_AUTH_USER';
export const INVALIDATE_AUTH_USER = 'INVALIDATE_AUTH_USER';
export const SET_AUTH_TOKEN = 'SET_AUTH_TOKEN';
