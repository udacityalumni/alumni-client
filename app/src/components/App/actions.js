import * as types from './constants';

export const appToggleNav = () => ({
  type: types.APP_ON_TOGGLE_NAV,
});

export const appSetMobile = (isMobile) => ({
  type: types.APP_SET_MOBILE,
  isMobile,
});

export const setSearchTerm = (term) => ({
  type: types.SET_SEARCH_TERM,
  term,
});

export const clearSearchTerm = () => ({
  type: types.CLEAR_SEARCH_TERM,
});

export const invalidateAuthUser = () => ({
  type: types.INVALIDATE_AUTH_USER,
});

export const setAuthUser = (user) => ({
  type: types.SET_AUTH_USER,
  user,
});

export const loadAuthToken = (token) => ({
  type: types.SET_AUTH_TOKEN,
  token,
});

export const loadPersistedAuthToken = () => (dispatch) => {
  const token = localStorage.getItem('auth_token');
  dispatch(
    loadAuthToken(token)
  );
};

export const setPersistentUser = (user) => (dispatch) => {
  localStorage.setItem('user', JSON.stringify(user));
  dispatch(
    setAuthUser(user)
  );
};

export const logoutUser = () => (dispatch) => {
  localStorage.setItem('user', null);
  dispatch(
    invalidateAuthUser()
  );
};

export const loadPersistedUser = () => (dispatch) => {
  const user = localStorage.getItem('user');
  if (user) {
    const parsedUser = JSON.parse(user);
    dispatch(
      setAuthUser(parsedUser)
    );
  } else {
    dispatch(
      invalidateAuthUser()
    );
  }
};
