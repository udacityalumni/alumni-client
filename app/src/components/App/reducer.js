import * as types from './constants';
import update from 'react-addons-update';

export const initialState = {
  user: null,
  isLoading: false,
  error: null,
  navLinks: [
    {
      url: '/careers',
      text: 'Careers',
    },
    {
      url: '/mentorship',
      text: 'Mentorship',
    },
    {
      url: '/meetups',
      text: 'Meetups',
    },
  ],
  navIsActive: false,
  isMobile: false,
  searchTerm: null,
  authToken: null,
};

const appReducer =
  (state = initialState, action) => {
    switch (action.type) {
      case types.SET_AUTH_TOKEN:
        return update(state, {
          authToken: {
            $set: action.token,
          },
        });
      case types.APP_SET_MOBILE:
        return update(state, {
          isMobile: {
            $set: action.isMobile,
          },
        });
      case types.APP_ON_TOGGLE_NAV:
        return update(state, {
          navIsActive: {
            $set: !state.navIsActive,
          },
        });
      case types.SET_SEARCH_TERM:
        return update(state, {
          searchTerm: {
            $set: action.term,
          },
        });
      case types.CLEAR_SEARCH_TERM:
        return update(state, {
          searchTerm: {
            $set: null,
          },
        });
      case types.SET_AUTH_USER:
        return update(state, {
          user: {
            $set: action.user,
          },
        });
      default:
        return state;
    }
  };


export default appReducer;
