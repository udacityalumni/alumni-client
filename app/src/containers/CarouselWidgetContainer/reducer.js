import * as types from './constants';
import update from 'react-addons-update';

export const initialState = {
  currentlyEditing: null,
  images: null,
};

const carouselWidgetReducer =
  (state = initialState, action) => {
    switch (action.type) {
      case types.CAROUSEL_SET_IMAGES:
        return update(state, {
          images: {
            $set: action.images,
          },
        });
      case types.CAROUSEL_SET_EDITING:
        return update(state, {
          currentlyEditing: {
            $set: action.index,
          },
        });
      case types.CAROUSEL_CANCEL_EDITING:
        return update(state, {
          currentlyEditing: {
            $set: null,
          },
        });
      case types.CAROUSEL_ADD_IMAGE:
        return update(state, {
          images: {
            $set: [...state.images, action.image],
          },
        });
      case types.CAROUSEL_REMOVE_IMAGE:
        return update(state, {
          images: {
            $set: [
              ...state.images.slice(0, action.index),
              ...state.images.slice(action.index + 1),
            ],
          },
        });
      case types.CAROUSEL_EDIT_IMAGE:
        return update(state, {
          images: {
            $set: [
              ...state.images.slice(0, action.index),
              action.image,
              ...state.images.slice(action.index + 1),
            ],
          },
          currentlyEditing: {
            $set: null,
          },
        });
      default:
        return state;
    }
  };

export default carouselWidgetReducer;
