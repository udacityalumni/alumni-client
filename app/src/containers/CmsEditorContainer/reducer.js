import * as types from './constants';
import update from 'react-addons-update';
import { editorStateFromRaw } from 'megadraft';

export const initialState = {
  article: {
    id: null,
    action: null,
  },
  editorState: editorStateFromRaw(null),
  editorTitle: null,
  preview: {
    isPreviewing: false,
    content: null,
    title: null,
  },
  isValid: false,
  isLoading: false,
  error: null,
  message: null,
  modal: {
    isShowing: false,
    status: 0,
    spotlighted: false,
    selectedTags: [],
  },
};

const previewReducer = (state = initialState.preview, action) => {
  switch (action.type) {
    case types.CMS_CLOSE_PREVIEW:
      return update(state, {
        isPreviewing: {
          $set: false,
        },
        markdown: {
          $set: null,
        },
        title: {
          $set: null,
        },
      });
    case types.CMS_SET_PREVIEW_STATE:
      return update(state, {
        isPreviewing: {
          $set: true,
        },
        content: {
          $set: action.markdown,
        },
        title: {
          $set: action.title,
        },
      });
    default: return state;
  }
};

const statusEnum = (status) => {
  switch (status) {
    case 'draft':
      return 0;
    case 'published':
      return 1;
    case 'archived':
      return 2;
    default: return 0;
  }
};

const modalReducer = (state = initialState.modal, action) => {
  switch (action.type) {
    case types.CMS_SET_STATE_FROM_ARTICLE:
      return update(state, {
        spotlighted: {
          $set: action.article.spotlighted,
        },
        status: {
          $set: statusEnum(action.article.status),
        },
        selectedTags: {
          $set: action.article.tags.map((tag) =>
            ({
              value: tag.tag,
              label: tag.tag,
            }),
          ),
        },
      });
    case types.CMS_TOGGLE_SPOTLIGHT:
      return update(state, {
        spotlighted: {
          $set: !state.spotlighted,
        },
      });
    case types.CMS_OPEN_MODAL:
      return update(state, {
        isShowing: {
          $set: true,
        },
      });
    case types.CMS_CLOSE_MODAL:
      return update(state, {
        isShowing: {
          $set: false,
        },
      });
    case types.CMS_SET_STATUS:
      return update(state, {
        status: {
          $set: action.status,
        },
      });
    case types.CMS_SET_SELECTED_TAGS:
      return update(state, {
        selectedTags: {
          $set: action.tags,
        },
      });
    default: return state;
  }
};

const cmsEditorReducer =
  (state = initialState, action) => {
    switch (action.type) {
      case types.SUBMIT_ARTICLE_INITIATION:
        return update(state, {
          isLoading: {
            $set: true,
          },
        });
      case types.SUBMIT_ARTICLE_FAILURE:
        return update(state, {
          isLoading: {
            $set: false,
          },
          error: {
            $set: action.error,
          },
        });
      case types.SUBMIT_ARTICLE_SUCCESS:
        return update(state, {
          isLoading: {
            $set: false,
          },
          message: {
            $set: action.message,
          },
        });
      case types.CLEAR_CMS_ERROR:
        return update(state, {
          error: {
            $set: null,
          },
        });
      case types.CLEAR_CMS_MESSAGE:
        return update(state, {
          message: {
            $set: null,
          },
        });
      case types.CMS_OPEN_MODAL:
        return update(state, {
          modal: {
            $set: modalReducer(state.modal, action),
          },
        });
      case types.CMS_CLOSE_MODAL:
        return update(state, {
          modal: {
            $set: modalReducer(state.modal, action),
          },
        });
      case types.CMS_SET_STATUS:
        return update(state, {
          modal: {
            $set: modalReducer(state.modal, action),
          },
        });
      case types.CMS_SET_SELECTED_TAGS:
        return update(state, {
          modal: {
            $set: modalReducer(state.modal, action),
          },
        });
      case types.CMS_TOGGLE_SPOTLIGHT:
        return update(state, {
          modal: {
            $set: modalReducer(state.modal, action),
          },
        });
      case types.CMS_SET_EDITOR_TITLE:
        return update(state, {
          editorTitle: {
            $set: action.title,
          },
          isValid: {
            $set: action.title !== '',
          },
        });
      case types.CMS_SET_EDITOR_STATE:
        return update(state, {
          editorState: {
            $set: action.state,
          },
        });
      case types.CMS_CLOSE_PREVIEW:
        return update(state, {
          preview: {
            $set: previewReducer(state.preview, action),
          },
        });
      case types.CMS_SET_PREVIEW_STATE:
        return update(state, {
          preview: {
            $set: previewReducer(state.preview, action),
          },
        });
      case types.CMS_SET_ARTICLE_ID:
        return update(state, {
          article: {
            id: {
              $set: parseInt(action.id, 10),
            },
            action: {
              $set: action.action,
            },
          },
        });
      case types.CMS_SET_STATE_FROM_ARTICLE:
        const rawContent = JSON.parse(action.article.json);
        return update(state, {
          isValid: {
            $set: true,
          },
          editorState: {
            $set: editorStateFromRaw(rawContent),
          },
          editorTitle: {
            $set: action.article.title,
          },
          modal: {
            $set: modalReducer(state.modal, action),
          },
        });
      default:
        return state;
    }
  };

export default cmsEditorReducer;
