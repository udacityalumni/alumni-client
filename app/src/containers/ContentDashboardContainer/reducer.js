import * as types from './constants';
import update from 'react-addons-update';

export const initialState = {
  articles: null,
  error: null,
  message: null,
  isLoading: false,
  isShowingModal: false,
  selectedArticleId: null,
};

const contentDashboardReducer =
  (state = initialState, action) => {
    switch (action.type) {
      case types.DASHBOARD_DELETE_ARTICLE_INITIATION:
        return update(state, {
          isLoading: {
            $set: true,
          },
        });
      case types.DASHBOARD_DELETE_ARTICLE_SUCCESS:
        return update(state, {
          isLoading: {
            $set: false,
          },
          message: {
            $set: action.message,
          },
          selectedArticleId: {
            $set: null,
          },
        });
      case types.DASHBOARD_DELETE_ARTICLE_FAILURE:
        return update(state, {
          isLoading: {
            $set: false,
          },
          error: {
            $set: action.error.message,
          },
          selectedArticleId: {
            $set: null,
          },
        });
      case types.DASHBOARD_TOGGLE_MODAL:
        return update(state, {
          isShowingModal: {
            $set: true,
          },
          selectedArticleId: {
            $set: action.id,
          },
        });
      case types.DASHBOARD_MODAL_CONFIRMATION:
        return update(state, {
          isShowingModal: {
            $set: false,
          },
          selectedArticleId: {
            $set: null,
          },
        });
      case types.DASHBOARD_MODAL_CANCEL:
        return update(state, {
          isShowingModal: {
            $set: false,
          },
          selectedArticleId: {
            $set: null,
          },
        });
      case types.DASHBOARD_ARTICLES_INITIATION:
        return update(state, {
          isLoading: {
            $set: true,
          },
        });
      case types.DASHBOARD_ARTICLES_SUCCESS:
        return update(state, {
          isLoading: {
            $set: false,
          },
          articles: {
            $set: action.articles,
          },
        });
      case types.DASHBOARD_ARTICLES_FAILURE:
        return update(state, {
          isLoading: {
            $set: false,
          },
          error: {
            $set: action.error,
          },
        });
      case types.CLEAR_DASHBOARD_ERROR:
        return update(state, {
          error: {
            $set: null,
          },
        });
      case types.CLEAR_DASHBOARD_MESSAGE:
        return update(state, {
          message: {
            $set: null,
          },
        });
      default:
        return state;
    }
  };

export default contentDashboardReducer;
