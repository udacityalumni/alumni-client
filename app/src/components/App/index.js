import React, { Component, PropTypes } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import * as actionCreators from './actions';
import App from 'grommet-udacity/components/App';
import { MobileNav, LogoImage, Navbar } from 'components';
import { updatePageTitle, getTitleFromRoute } from 'utils/a11y';
import Header from 'grommet-udacity/components/Header';
import Title from 'grommet-udacity/components/Title';
import Anchor from 'grommet-udacity/components/Anchor';
import MenuIcon from 'grommet-udacity/components/icons/base/Menu';

class Main extends Component {
  constructor() {
    super();
    this.handleToggleNav = this.handleToggleNav.bind(this);
    this.handleSetMobile = this.handleSetMobile.bind(this);
    this.handleSearch = this.handleSearch.bind(this);
    this.loadUser = this.loadUser.bind(this);
  }
  componentDidMount() {
    this.loadUser();
    const {
      pathname,
    } = this.props.location;
    updatePageTitle(getTitleFromRoute(pathname));
    this.handleSetMobile();
    if (window) {
      window.addEventListener('resize', this.handleSetMobile);
    }
  }
  componentWillReceiveProps(newProps) {
    const {
      pathname,
    } = this.props.location;
    const newPathname = newProps.location.pathname;
    if (newPathname !== pathname) {
      updatePageTitle(getTitleFromRoute(newPathname));
    }
  }
  componentWillUnmount() {
    if (window) {
      window.removeEventListener('resize', this.handleSetMobile);
    }
  }
  loadUser() {
    const {
      loadPersistedUser,
      loadPersistedAuthToken,
    } = this.props.actions;
    loadPersistedUser();
    loadPersistedAuthToken();
  }
  handleSetMobile() {
    const isMobile = window.innerWidth <= 768;
    const {
      appSetMobile,
    } = this.props.actions;
    appSetMobile(isMobile);
  }
  handleToggleNav() {
    const {
      appToggleNav,
    } = this.props.actions;
    appToggleNav();
  }
  handleSearch(e) {
    const {
      router,
    } = this.context;
    const {
      setSearchTerm,
      clearSearchTerm,
    } = this.props.actions;
    if (e.target.value) {
      setSearchTerm(e.target.value);
      router.push('/search');
    } else {
      clearSearchTerm();
    }
  }
  render() {
    const {
      user,
      isMobile,
      navIsActive,
      navLinks,
    } = this.props;
    return (
      <App centered={false}>
        {!isMobile ?
          <Navbar user={user} onSearch={this.handleSearch} />
        :
          <MobileNav
            user={user}
            navActive={navIsActive}
            onToggleNav={this.handleToggleNav}
            navLinks={navLinks}
          >
            <Header
              direction="row"
              justify="between"
              large
              pad={{ horizontal: 'medium', between: 'small' }}
            >
              {!navIsActive &&
                <Title a11yTitle="Go Home">
                  <Anchor href="/">
                    <LogoImage />
                  </Anchor>
                </Title>
              }
              <Title onClick={this.handleToggleNav} a11yTitle="Open Menu Right">
                <MenuIcon colorIndex="brand" size="medium" type="control" />
              </Title>
            </Header>
          </MobileNav>
        }
        {React.cloneElement(this.props.children, this.props)}
      </App>
    );
  }
}

Main.propTypes = {
  children: React.children,
  location: PropTypes.object.isRequired,
  user: PropTypes.object,
  isMobile: PropTypes.bool.isRequired,
  navIsActive: PropTypes.bool.isRequired,
  navLinks: PropTypes.array.isRequired,
  actions: PropTypes.object.isRequired,
};

Main.contextTypes = {
  router: PropTypes.object.isRequired,
};

// Map the global state to global props here.
// See: https://egghead.io/lessons/javascript-redux-generating-containers-with-connect-from-react-redux-visibletodolist
// mapStateToProps :: {State} -> {Action}
const mapStateToProps = (state) => ({
  user: state.app.user,
  navIsActive: state.app.navIsActive,
  isMobile: state.app.isMobile,
  navLinks: state.app.navLinks,
});

// Map the dispatch and bind the action creators.
// See: http://redux.js.org/docs/api/bindActionCreators.html
// mapDispatchToProps :: Dispatch Func -> {Actions}
const mapDispatchToProps = (dispatch) => ({
  actions: bindActionCreators(
    actionCreators,
    dispatch
  ),
});

// Use connect both here and in your components.
// See: https://egghead.io/lessons/javascript-redux-generating-containers-with-connect-from-react-redux-visibletodolist
const ConnectedApp = connect(
  mapStateToProps,
  mapDispatchToProps
)(Main);

export default ConnectedApp;
