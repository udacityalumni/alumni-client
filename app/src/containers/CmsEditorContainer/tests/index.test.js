import CmsEditor from '../index';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { initialState as cmsEditorContainer } from '../reducer';
import { initialState as app } from '../../../components/App/reducer';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('<CmsEditor />', () => {
  it('should render with default props', () => {
    const store = mockStore({ cmsEditorContainer, app });
    const wrapper = shallow(
      <CmsEditor store={store} />
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
