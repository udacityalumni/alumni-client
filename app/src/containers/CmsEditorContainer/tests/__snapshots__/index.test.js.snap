exports[`<CmsEditor /> should render with default props 1`] = `
<Apollo(Apollo(CmsEditorContainer))
  action={null}
  actions={
    Object {
      "clearCmsError": [Function anonymous],
      "clearCmsMessage": [Function anonymous],
      "cmsCloseModal": [Function anonymous],
      "cmsClosePreview": [Function anonymous],
      "cmsOpenModal": [Function anonymous],
      "cmsSetArticleId": [Function anonymous],
      "cmsSetEditorState": [Function anonymous],
      "cmsSetEditorTitle": [Function anonymous],
      "cmsSetPreviewState": [Function anonymous],
      "cmsSetSelectedTags": [Function anonymous],
      "cmsSetStateFromArticle": [Function anonymous],
      "cmsSetStatus": [Function anonymous],
      "cmsToggleSpotlight": [Function anonymous],
      "handleClearingToast": [Function anonymous],
      "submitArticleFailure": [Function anonymous],
      "submitArticleInitiation": [Function anonymous],
      "submitArticleRequest": [Function anonymous],
      "submitArticleSucces": [Function anonymous]
    }
  }
  articleId={null}
  editorState={
    EditorState {
      "_immutable": Object {
        "allowUndo": true,
        "currentContent": Object {
          "blockMap": Object {
            "305pe": Object {
              "characterList": Array [],
              "data": Object {},
              "depth": 0,
              "key": "305pe",
              "text": "",
              "type": "unstyled"
            }
          },
          "selectionAfter": Object {
            "anchorKey": "305pe",
            "anchorOffset": 0,
            "focusKey": "305pe",
            "focusOffset": 0,
            "hasFocus": false,
            "isBackward": false
          },
          "selectionBefore": Object {
            "anchorKey": "305pe",
            "anchorOffset": 0,
            "focusKey": "305pe",
            "focusOffset": 0,
            "hasFocus": false,
            "isBackward": false
          }
        },
        "decorator": CompositeDraftDecorator {
          "_decorators": Array [
            Object {
              "component": [Function Link],
              "strategy": [Function findLinkEntities]
            }
          ]
        },
        "directionMap": Object {
          "305pe": "LTR"
        },
        "forceSelection": false,
        "inCompositionMode": false,
        "inlineStyleOverride": null,
        "lastChangeType": null,
        "nativelyRenderedContent": null,
        "redoStack": Array [],
        "selection": Object {
          "anchorKey": "305pe",
          "anchorOffset": 0,
          "focusKey": "305pe",
          "focusOffset": 0,
          "hasFocus": false,
          "isBackward": false
        },
        "treeMap": Object {
          "305pe": Array [
            Object {
              "decoratorKey": null,
              "end": 0,
              "leaves": Array [
                Object {
                  "end": 0,
                  "start": 0
                }
              ],
              "start": 0
            }
          ]
        },
        "undoStack": Array []
      }
    }
  }
  editorTitle={null}
  isSubmitting={false}
  isValid={false}
  message={null}
  modal={
    Object {
      "isShowing": false,
      "selectedTags": Array [],
      "spotlighted": false,
      "status": 0
    }
  }
  preview={
    Object {
      "content": null,
      "isPreviewing": false,
      "title": null
    }
  }
  store={
    Object {
      "clearActions": [Function clearActions],
      "dispatch": [Function anonymous],
      "getActions": [Function getActions],
      "getState": [Function getState],
      "replaceReducer": [Function replaceReducer],
      "subscribe": [Function subscribe]
    }
  }
  submissionError={null}
  user={null} />
`;
